File: packages/hmd-core/src/graph.ts
/**
 * The resolved graph and the reverse edge map (HMD-0021 §9, §10).
 *
 * Backlinks are a query against an index the workspace already maintains, not
 * a scan: "what links here" is the cheapest question a knowledge base can be
 * asked, and it should stay that way as the tree grows.
 */

import type { BacklinkEntry } from "./ir.js";
import type { Workspace } from "./workspace.js";

export interface GraphNode {
  path: string;
  namespace: string;
  tags: readonly string[];
  headings: number;
  autodiscovery: boolean;
}

export interface GraphEdge {
  source: string;
  target: string | null;
  kind: "link" | "embed";
  raw: string;
  outcome: string;
  phase: number | null;
  fragment: string | null;
  line: number;
  column: number;
}

export interface Graph {
  nodes: GraphNode[];
  edges: GraphEdge[];
}

export function buildGraph(workspace: Workspace): Graph {
  const nodes: GraphNode[] = [];
  const edges: GraphEdge[] = [];

  for (const path of workspace.pages()) {
    const document = workspace.documents.get(path);
    if (document === undefined) continue;
    const cut = path.lastIndexOf("/");
    nodes.push({
      path,
      namespace: cut === -1 ? "/" : path.slice(0, cut),
      tags: document.card.tags,
      headings: document.headings.length,
      autodiscovery: workspace.autodiscoveryEnabled(path),
    });

    for (const link of document.links) {
      const result = workspace.resolve(path, link.pageRef);
      edges.push({
        source: path,
        target: result.path,
        kind: link.isEmbed ? "embed" : "link",
        raw: link.raw,
        outcome: result.outcome,
        phase: result.phase,
        fragment: link.fragment,
        line: link.span.line,
        column: link.span.column,
      });
    }
  }

  nodes.sort((a, b) => (a.path < b.path ? -1 : a.path > b.path ? 1 : 0));
  edges.sort(
    (a, b) =>
      (a.source < b.source ? -1 : a.source > b.source ? 1 : 0) ||
      a.line - b.line ||
      a.column - b.column,
  );
  return { nodes, edges };
}

/** Every card with a resolved edge into `path`, with a one-line snippet. */
export function backlinks(workspace: Workspace, path: string): BacklinkEntry[] {
  const out: BacklinkEntry[] = [];

  for (const source of workspace.pages()) {
    if (source === path) continue;
    const document = workspace.documents.get(source);
    if (document === undefined) continue;
    const lines = document.text.split("\n");

    for (const link of document.links) {
      const result = workspace.resolve(source, link.pageRef);
      if (result.path !== path) continue;
      out.push({
        path: source,
        line: link.span.line,
        kind: link.isEmbed ? "embed" : "link",
        snippet: (lines[link.span.line - 1] ?? "").trim().slice(0, 200),
      });
    }
  }

  out.sort((a, b) => (a.path < b.path ? -1 : a.path > b.path ? 1 : 0) || a.line - b.line);
  return out;
}

File: tools/hmd/src/hyper_markdown/graph.py
"""Resolved-graph assembly and JSON serialization (HMD-0001 §7)."""

from __future__ import annotations

import json

from .resolve import Outcome, Workspace


def build(workspace: Workspace) -> dict:
    """Nodes with frontmatter and headings, edges with kind and source span."""
    nodes = []
    edges = []

    for path in workspace.pages():
        document = workspace.documents[path]
        rel = workspace.rel(path)
        table = workspace.import_table(path)
        nodes.append(
            {
                "path": rel,
                "namespace": path.parent.relative_to(workspace.root).as_posix() or "/",
                "tags": list(document.card.tags),
                "frontmatter": _plain(document.frontmatter),
                "headings": [
                    {"level": h.level, "text": h.text, "slug": h.slug, "line": h.span.line}
                    for h in document.headings
                ],
                "anchors": [a.block_id for a in document.anchors],
                "autodiscovery": workspace.autodiscovery_enabled(path),
                "imports": {
                    "bindings": {k: workspace.rel(v) for k, v in sorted(table.bindings.items())},
                    "search_paths": [workspace.rel(p) if p != workspace.root else "/" for p in table.search_paths],
                },
            }
        )

        for link in document.links:
            result = workspace.resolve(path, link.page_ref)
            edges.append(
                {
                    "source": rel,
                    "target": workspace.rel(result.path) if result.path else None,
                    "kind": "embed" if link.is_embed else "link",
                    "raw": link.raw,
                    "outcome": result.outcome.value,
                    "phase": result.phase,
                    "fragment": link.fragment,
                    "line": link.span.line,
                    "column": link.span.column,
                }
            )

    nodes.sort(key=lambda n: n["path"])
    edges.sort(key=lambda e: (e["source"], e["line"], e["column"]))
    return {"root": str(workspace.root), "nodes": nodes, "edges": edges}


def to_json(workspace: Workspace) -> str:
    return json.dumps(build(workspace), indent=2, sort_keys=True)


def _plain(value):
    """Coerce YAML scalars to JSON-serializable values."""
    if isinstance(value, dict):
        return {str(k): _plain(v) for k, v in value.items()}
    if isinstance(value, (list, tuple)):
        return [_plain(v) for v in value]
    if isinstance(value, (str, int, float, bool)) or value is None:
        return value
    return str(value)


__all__ = ["build", "to_json", "Outcome"]
